IHP/RouterSupport.hs

```diff
@@ -100,8 +99,145 @@ class HasPath controller => CanRoute controller where
+    where
+        findInQuery _ [] = Nothing
+        findInQuery field ( (k,v) : rest ) = if field == k then v else findInQuery field rest
+
+data TypedAutoRouteError
+    = BadType
+    | TooFewArguments
+    | NotMatched
+    | NoConstructorMatched
+    deriving (Show)
+
+-- | Given a constructor and a parsed query string, attempt to construct a value of the constructor's type.
+-- For example, given the controller
+--
+-- @data MyController = MyAction { textArg :: Text, intArg :: Int }@
+--
+-- this function will receive a representation of the @MyAction@ constructor as well as some query string
+-- @[("textArg", "some text"), ("intArg", "123")]@.
+--
+-- By iterating through the query and attempting to match the type of each constructor argument
+-- with some transformation of the query string, we attempt to call @MyAction@.
+applyConstr :: Data controller => Constr -> Query -> Either TypedAutoRouteError controller
+applyConstr constructor query = let
+
+    -- |Given some query item (key, optional value), try to parse into the current expected type
+    -- by iterating through the available parse functions.
+    attemptToParseArg :: forall d. (Data d) => (ByteString, Maybe ByteString) -> [Maybe ByteString -> Either TypedAutoRouteError d] -> State.StateT Query (Either TypedAutoRouteError) d
+    attemptToParseArg _ [] = State.lift (Left NoConstructorMatched)
+    attemptToParseArg queryParam@(k, v) (parseFunc:restFuncs) = case parseFunc v of
+            Right result -> pure result
+            -- BadType will be returned if, for example, a text is passed to a query parameter typed as int.
+            Left BadType -> State.lift (Left BadType)
+            -- otherwise, safe to assume the match just failed, so recurse on the rest of the functions and try to find one that matches.
+            Left _ -> attemptToParseArg queryParam restFuncs
+
+    -- | Attempt to parse the current expected type, and return its value.
+    -- For the example @MyController@ this is called twice by @fromConstrM@.
+    -- Once, it is called for @textArg@ where @d :: Text@. Then it is called
+    -- for @intArg@ with @d ::: Int@. With both of these values parsed from the query string,
+    -- the controller action is able to be created.
+    nextField :: forall d. (Data d) => State.StateT Query (Either TypedAutoRouteError) d
+    nextField = do
+            queryParams <- State.get
+            case queryParams of
+                [] -> State.lift (Left TooFewArguments)
+                (p@(key, value):rest) -> do
+                    State.put rest
+                    attemptToParseArg p parseFuncs
+
+
+   in case State.runStateT (fromConstrM nextField constructor) (querySortedByFields query constructor) of
+        Right (x, []) -> pure x
+        Right (_) -> Left TooFewArguments
+        Left e -> Left e  -- runtime type error
+
+
 class Data controller => AutoRoute controller where
-    {-# INLINE autoRoute #-}
+    -- | DEPRECATED.
```

Comment:
Why is this deprecated? Cannot we just replace the existing `autoRoute` with the new `typedAutoRoute`?

Reply:
Yeah we can replace it, that works too. Wasn't sure how IHP handles breaking changes. I guess it's a good time since it's still beta!

Reply:
What breaking changes you have in mind? While looking over the code I got the feeling that for 90% of people nothing will break.

Reply:
Removing the `parseArgument` functions since they're no longer used would cause compiles to break in all the places where `parseArgument = parseTextArgument` is set in the AutoRoute instances. Other than that everything should work the same

Reply:
👍  I think this is worth breaking, it's a price worth paying for progress :)